### ISSUE TITLE
sonyAPILib.cs - function HttpGet(string Url) - HttpWebRequest req.Host variable is set wrong

### ISSUE THREAD

**Issue by mary-17** [KHerron/SonyAPILib#10]
I am testing the SonyAPILib together with my Sony STR-DN1040 AV Receiver.

After I managed the registration process with the STR-DN1040, I was able to read the [RemoteCommandList](https://github.com/KHerron/SonyAPILib/wiki/1.3.1-Example-Remote-Command-List) XML file.
But there is one bug in function HttpGet(string Url) in file sonyAPILib.cs: the port of the HttpWebRequest req.Host variable is set wrong. The port must be parsed from the Url string parameter and not from this.Device_Port.

The description.xml of the STR-DN1040 must be read via port 8080, the ActionList.xml and /cers/getRemoteCommandList must be read via port 50001.

I made following hack to read the STR-DN1040's command list:

`//req.Host = this.Device_IP_Address + ":" + this.Device_Port;`
`req.Host = this.Device_IP_Address + ":50001";`

I used buildFromDocument(string dDoc, string fPath) to generate my SonyDevice via your DLNA_Control_Panel GUI button "Build Device from Document".

**Comment by KHerron** [comment]
> I am testing the SonyAPILib together with my Sony STR-DN1040 AV Receiver.
> 
> After I managed the registration process with the STR-DN1040, I was able to read the RemoteCommandList XML file.
> But there is one bug in function HttpGet(string Url) in file sonyAPILib.cs: the port of the HttpWebRequest req.Host variable is set wrong. The port must be parsed from the Url string parameter and not from this.Device_Port.
> 

Can you please tell me what your were trying to do when you encountered the error.

Port Numbers have been an issue before, as it changes according to device type and generation. Also the HttpGet function is used for other commands, that use different ports as well. However, I was not aware of any issues, as I have a STR-DN840. 

So keep in mind that even though your hack fixed this issue, Hard-Coding the port to "50001" may cause other functions to not work...

Can you forward me a copy of your Device XML file? 
Or can you tell me what the <getRemoteCommandList> property is set to?

**Comment by KHerron** [comment]
After further investigation, I think I have accidentally lost some coding that should have been placed here to help correct the Porting issues. I must have lost them during my last update and was not aware. This is probably why I am not having the same issues here.

Basically there is an evaluation of the Url that is sent to see if in includes a port number. If it does, it uses it instead of the devices port number. I will try to post an update here.

The first few lines should look something like this:
```
  private string HttpGet(string Url)
            {
                int getPort = this.Device_Port;
                Uri getUri = new Uri(Url);
                if(getUri.Port != getPort)
                {
                    getPort = getUri.Port;
                }
                _Log.writetolog("Creating HttpWebRequest to URL: " + Url, true);
                HttpWebRequest req = (HttpWebRequest)WebRequest.Create(Url);
                req.KeepAlive = true;
                // Set our default header Info
                _Log.writetolog("Setting Header Information: " + req.Host.ToString(), false);
                req.Host = this.Device_IP_Address + ":" + getPort;
```

**Comment by mary-17** [comment]
Thank you for your response.

> Can you please tell me what your were trying to do when you encountered the error.

Yes.

1) First of all I created a Visual Studio solution with DLNA_Control_Panel as start project and SonyAPILib. I allways start DLNA_Control_Panel in debug mode via Visual Studio.


2) I tried button LocateBut which calls mySonyLib.Locator.locateDevices(). But my receiver and my NAS were not found (only my Smart TV, my Chromecast and my Router).
   --> Shall I create an other issue/bug report for this? 
   This bug was not my priority, because I only want to develop some tools for my own smart home and learn about UPnP/DLNA.


3) So I typed the link to the description.xml of the receiver to TextBox bDocURL ("http://192.168.100.5:8080/description.xml") and clicked BuildBut. In my home network all devices have a static IP, so the use of curDev.buildFromDocument is sufficient for me. At this time I didn't done the registration procedure of my Sony receiver yet, so the HttpGet call inside get_remote_command_list ends in an exception:

08.08.2016 15:15:15 - Document Found
08.08.2016 15:15:15 - Action List Found.
08.08.2016 15:15:15 - Device has a registration Mode of: 1
08.08.2016 15:15:15 -  is Retrieving Generation:1 Remote Command List
08.08.2016 15:15:15 - Creating HttpWebRequest to URL: http://192.168.100.5:50001/cers/getRemoteCommandList
08.08.2016 15:15:15 - Setting Header Information: 192.168.100.5:50001
08.08.2016 15:15:15 - Creating Web Request Response
08.08.2016 15:15:15 - There was an error during the Web Request or Response! System.Net.WebException: Die zugrunde liegende Verbindung wurde geschlossen: Die Verbindung wurde unerwartet getrennt..
   bei System.Net.HttpWebRequest.GetResponse()
   bei SonyAPILib.SonyAPI_Lib.SonyDevice.HttpGet(String Url) in ..\SonyAPILib\sonyAPILib.cs:Zeile 533.
08.08.2016 15:15:15 - Retrieve Command List was Successful

   The return value of HttpGet in variable cmdList is "false : System.Net.WebException ....", so the else-path "Retrieve Command List was NOT successful" isn't called. So I have an other exception: 
( German: Ausnahme=Exception; Zeile=Line )

Eine Ausnahme (erste Chance) des Typs "System.Xml.XmlException" ist in System.Xml.dll aufgetreten.
System.Xml.XmlException: Ungültige Daten auf Stammebene. Zeile 1, Position 1.
   bei System.Xml.XmlTextReaderImpl.Throw(Exception e)
   bei System.Xml.XmlTextReaderImpl.ParseRootLevelWhitespace()
   bei System.Xml.XmlTextReaderImpl.ParseDocumentContent()
   bei System.Xml.XmlReader.MoveToContent()
   bei System.Data.DataSet.ReadXml(XmlReader reader, Boolean denyResolving)
   bei System.Data.DataSet.ReadXml(XmlReader reader, XmlReadMode mode, Boolean denyResolving)
   bei SonyAPILib.SonyAPI_Lib.SonyDevice.get_remote_command_list() in ..\SonyAPILib\sonyAPILib.cs:Zeile 470.
   bei SonyAPILib.SonyAPI_Lib.SonyDevice.buildFromDocument(String dDoc, String fPath) in ..\SonyAPILib\sonyAPILib.cs:Zeile 1259.
   bei SonyAPILib.SonyAPI_Lib.SonyDevice.buildFromDocument(Uri dPath) in ..\SonyAPILib\sonyAPILib.cs:Zeile 1215.
   bei DLNA_Control_Panel.Form1.BuildBut_Click(Object sender, EventArgs e) in ..\DLNA_Control_Panel\Form1.cs:Zeile 92.  // ==> curDev.buildFromDocument(new Uri(bDocURL.Text));

   --> Shall I create an other issue/bug report for this? I haven't done it, because it may only occur without the registration procedure of the receiver.


4) Then I commented out get_remote_command_list() call in buildFromDocument() of sonyAPILib.cs. Now I have an other exception in Form1.UpdateIRCC() because curDev.Commands is null:
( German: Der Wert darf nicht NULL sein. = The Value cannot be NULL. )

System.ArgumentNullException: Der Wert darf nicht NULL sein.
Parametername: source
   bei System.Linq.Enumerable.Count[TSource](IEnumerable`1 source)
   bei DLNA_Control_Panel.Form1.UpdateIRCC() in ..\DLNA_Control_Panel\Form1.cs:Zeile 181.
   bei DLNA_Control_Panel.Form1.UpdateDevice() in ..\DLNA_Control_Panel\Form1.cs:Zeile 164.
   bei DLNA_Control_Panel.Form1.BuildBut_Click(Object sender, EventArgs e) in ..\DLNA_Control_Panel\Form1.cs:Zeile 95.
   bei System.Linq.Enumerable.Count[TSource](IEnumerable`1 source)
   bei DLNA_Control_Panel.Form1.UpdateIRCC() in ..\DLNA_Control_Panel\Form1.cs:Zeile 181.
   bei DLNA_Control_Panel.Form1.UpdateDevice() in ..\DLNA_Control_Panel\Form1.cs:Zeile 164.
   bei DLNA_Control_Panel.Form1.BuildBut_Click(Object sender, EventArgs e) in ..\DLNA_Control_Panel\Form1.cs:Zeile 95.

   --> Shall I create an other issue/bug report for this? I haven't done it, because it may only occur without the registration procedure of the receiver.


5) I made further changes in the code regarding exception handling, performed the registration procedure with my Sony receiver and finally found the port issue in the HttpGet.


> So keep in mind that even though your hack fixed this issue, Hard-Coding the port to "50001" may cause other functions to not work...

Yes, of cause.

> Can you forward me a copy of your Device XML file? 

Here my decription.xml:
[description.zip](https://github.com/KHerron/SonyAPILib/files/406968/description.zip)

And here the ActionList.xml:
[ActionList.zip](https://github.com/KHerron/SonyAPILib/files/406972/ActionList.zip)


I will test your patch later.

**Comment by mary-17** [comment]
[Patch for 5.2 Issue 10](https://github.com/KHerron/SonyAPILib/commit/db6e2bcb182084c023ed3eacede138dbfac020cf) is working. Thanks.

Now to my point 2):

> Are you wired or using WiFi?

My PC was wired to my router. The receiver was also wired to the router. And the NAS was wired to a switch. --> locateDevices() does not find my NAS and my receiver.
Now my PC, my receiver and my NAS are all wired to a switch which is wired to the router. --> locateDevices() finds all of my UPnP devices.

**Comment by KHerron** [comment]
For some reason it sounds like your router is not passing the SSDP through to all the ports....
Not sure if this would be an API bug or some type of issue with your router.

I was having intermitting issues when I tried to use the Wifi on my STR-DB840. Once I ran a cable to it and connected it to my switch, I no longer have any issues with the LocateDevices() function. It could also be a timeout issue, I am only giving the function 15000ms to receive all responses. This may need to be increase to allow for more time.....

I will have to play and ponder this. I may even allow users to specify the time limit by changing the function to LocateDevices(TimeOut)....
